vignettes/forcats.Rmd

```diff
@@ -44,16 +44,55 @@ To do so, we can use the function `fct_infreq()`:
 
 ```{r fct-infreq-hair}
 #| fig.alt: >
-#|   The A bar chart of hair color, now ordered so that the least
+#|   The A bar chart of hair color, now ordered so that generally the least
 #|   frequent colours come first and the most frequent colors come last.
 #|   This makes it easy to see that the most common hair color is none 
-#|   (~35), followed by brown (~18), then black (~12).
-ggplot(starwars, aes(x = fct_infreq(hair_color))) + 
+#|   (~35), followed by brown (~18), then black (~12). Surprisingly,
+#|   NAs are at the top of the graph, even though there are ~5 NAs and other
+#|   colors have smaller values.
+ggplot(starwars, aes(y = fct_infreq(hair_color))) + 
   geom_bar() + 
-  coord_flip()
+  labs(y = "Hair color")
 ```
 
-Note that `fct_infreq()` it automatically puts NA at the top, even though that doesn't have the smallest number of entries.
+It's a little surprising that the `NA` bar isn't ordered by frequency.
+To understand why, we need to make a brief digression to discuss implicit vs explicit `NA`s in factors.
+
+## Implicit vs explicit missing values
+
+There are two ways to representing a missing value in a factor which we call implicit and :
+
+-   An **implicit** missing value does not appear in the levels and `is.na()` reports it as missing. This is the usual way to encode a missing value in a factor:
+
+    ```{r}
+    f <- factor(c("x", "y", NA))
+    levels(f)
+    is.na(f)
+    ```
+
+-   An **explicit** missing appears in the levels and `is.na()` does not report it as missing. This requires a little more work to create:
+
+    ```{r}
+    f <- factor(c("x", "y", NA), exclude = NULL)
+    levels(f)
+    is.na(f)
+    ```
+
+Implicit `NA`s tend to be best for data analysis, since `is.na()` works as you'd expect.
+Explicit `NA`s are often better for display since you can control where they appear.
```

Comment:
Here is another example with `table()`. I guess it is situational as to whether or not you want missing values in the counts at all

``` r
library(forcats)

x <- factor(c("a", NA, "c", "c", NA, "d"))

# implicit NA not reported by default
table(x)
#> x
#> a c d 
#> 1 2 1

# report it
table(x, useNA = "ifany")
#> x
#>    a    c    d <NA> 
#>    1    2    1    2

# explicit NA always reported
table(fct_explicit_na(x, na_level = "NA"))
#> 
#>  a  c  d NA 
#>  1  2  1  2
```

<sup>Created on 2023-01-03 with [reprex v2.0.2.9000](https://reprex.tidyverse.org)</sup>

Reply:
I'm remembering that when I designed `hardhat::weighted_table()` I avoided the need to add the `useNA` argument altogether by asserting that `weighted_table()` will use the _exact_ set of levels returned by `levels()` and nothing else, which ended up having some nice properties

<img width="934" alt="Screen Shot 2023-01-03 at 4 34 57 PM" src="https://user-images.githubusercontent.com/19150088/210445243-551ab46b-3564-4038-a1e1-f9a51aa492ac.png">

Reply:
I've always found it weird that `table()` drops missing values since pretty much every other summary preserves them unless you explicitly ask to remove.